packages/main/src/plugin/proxy.ts

```diff
@@ -98,36 +104,51 @@ export class Proxy {
     this.configurationRegistry.registerConfigurations([proxyConfigurationNode]);
 
     // be notified when configuration is updated
-    this.configurationRegistry.onDidChangeConfiguration(e => {
+    this.configurationRegistry.onDidChangeConfiguration(async e => {
       if (e.key.startsWith('proxy')) {
-        this.updateFromConfiguration();
+        await this.updateFromConfiguration();
       }
     });
 
     // read initial value
-    this.updateFromConfiguration();
+    await this.updateFromConfiguration();
     this.overrideFetch();
   }
 
-  updateFromConfiguration(): void {
+  async updateFromConfiguration(): Promise<void> {
     // read value from the configuration
     const proxyConfiguration = this.configurationRegistry.getConfiguration('proxy');
-    const httpProxy = proxyConfiguration.get<string>('http');
-    const httpsProxy = proxyConfiguration.get<string>('https');
-    const noProxy = proxyConfiguration.get<string>('no');
-    const isEnabled = proxyConfiguration.get<boolean>('enabled') ?? false;
-
-    this.proxySettings = {
-      httpProxy: ensureURL(httpProxy),
-      httpsProxy: ensureURL(httpsProxy),
-      noProxy,
-    };
-    this.proxyState = isEnabled;
+    const isEnabled = proxyConfiguration.get<unknown>('enabled');
+    let proxyState: ProxyState = ProxyState.PROXY_SYSTEM;
+    if (typeof isEnabled === 'boolean') {
+      proxyState = isEnabled ? ProxyState.PROXY_MANUAL : ProxyState.PROXY_SYSTEM;
+    } else if (typeof isEnabled === 'number') {
+      proxyState = isEnabled as ProxyState;
+    }
+    this.proxyState = proxyState;
+    if (this.proxyState === ProxyState.PROXY_MANUAL) {
+      const httpProxy = ensureURL(proxyConfiguration.get<string>('http'));
+      const httpsProxy = ensureURL(proxyConfiguration.get<string>('https'));
+      const noProxy = proxyConfiguration.get<string>('no');
+      this.proxySettings = {
+        httpProxy,
+        httpsProxy,
+        noProxy,
+      };
+    } else if (this.proxyState === ProxyState.PROXY_SYSTEM) {
+      this.proxySettings = await getProxySettingsFromSystem(this);
+    } else {
+      this.proxySettings = {} as ProxySettings;
```

Comment:
why do we use an empty object as ProxySettings

I would expect that it's undefined

Reply:
because we expect some fields from a ProxySettings object if defined